text/xxxx-metrics-correlation.md

```diff
@@ -0,0 +1,140 @@
+        for item in batch:
+            success = processor.process_item(item)
+            # This records an increment of 1
+            metrics.incr("processor.item_processed", tags={"success": success})
+        # This records a gauge
+        metrics.gauge("processor.peak_memory_usage", processor.peak_memory_usage)
+```
+
+Each metric locally "aggregates" into something that represents a gauge and is persisted with
+the closest span.  In the above case the following span gets recorded assuming a batch size of 5
+where 3 succeed and two fail, the following measurements might be associated:
+
+```json
+{
+    "span_id": "deadbeef",
+    "op": "timing_measurement",
+    "measurements": {
+        "d:processor.process_batch@millisecond": [
+            {
+                "min": 421.0,
+                "max": 421.0,
+                "count": 1,
+                "sum": 421.0
+            }
+        ],
+        "c:processor.item_processed": [
+            {
+                "min": 1,
+                "max": 1,
+                "count": 3,
+                "sum": 3,
+                "tags": {"success": True}
+            },
+            {
+                "min": 1,
+                "max": 1,
+                "count": 2,
+                "sum": 2,
+                "tags": {"success": False}
+            }
+        ],
+        "g:processor.peak_memory_usage@megabyte": {
+            {
+                "min": 42.0,
+                "max": 421.0,
+                "count": 1,
+                "sum": 421.0,
+            }
+        }
+    }
+}
+```
+
+Note that per-span all metrics (other than `sets` which are not yet accounted for) are
+stored as gauges.  That means that despite the fact that `incr` was called 5 times, only
+two gauges are stored.  The values of `min` and `max` are then useful for finding spans
+within the search radius.  Likewise tags are stored per metric and might diverge from
+the span tags.
+
+In a way this implies that there are two aggregators: a global aggregator and a per-span
+local gauge level minimal aggregator.
+
+# Intended Correlations
+
+The following correlations are useful for metrics to span queries:
+
+## Spans linked Timings
+
+When a code block is timed with `metrics.timing` (or potentially a span is named with the
+`metric` parameter) it emits a distribution as timing.  That also binds and creates a span
+and attached that metric directly as measurement.  In that case the tags for metrics
+might also have to be explicitly recorded with `metric_tags` as parameter.
+
+```python
+with start_span(description="something human readable", metric="my.span"):
+    pass
+```
+
+The corresponding low-level API (`metrics.timing`) is supposed to automatically create a
+span if tracing is enabled and no explict value is provided.  The following two examples
+are equivalent:
+
+```python
+with metrics.timing("foo"):
+    pass
+
+with start_span(op="timing_measurement", metric="foo"):
```

Comment:
how does this work when we are using an otel API under the hood? We'll have to add logic in otel span processors I guess?

Reply:
Yes, that is an open question I'm not fully sure about yet. OTel Metrics themselves do not suggest any span based metrics changes. They are doing the correlations OOB with exemplars: https://opentelemetry.io/docs/specs/otel/metrics/data-model/#exemplars

But I do think the case of span + timed metric is very common and ideally we can find a way to funnel that out with otel span APIs somehow.